contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/azkaban/GobblinDistcpJob.java

```diff
@@ -0,0 +1,195 @@
+    private final Props props;
+
+    GobblinDistcpJob(String id, String source, String destination, Props props) {
+        super(id, PrefixedLogger.getLogger(AdminStoreSwapper.class.getName(), destination));
+        this.source = source;
+        this.destination = destination;
+        this.props = props;
+    }
+
+    public void run() throws Exception {
+        String cdnURL = pickCDN();
+
+        if (!cdnURL.isEmpty()) {
+            info("####################################");
+            info("              Distcp");
+            info("####################################");
+            info("Using CDN: " + cdnURL);
+            try {
+                String username = props.getString("env.USER", "unknownClient");
+                String pathPrefix = props.getString(VoldemortBuildAndPushJob.PUSH_CDN_PREFIX, "/tmp/VoldemortBnP/");
+                pathPrefix = pathPrefix.endsWith("/") ? pathPrefix : pathPrefix + "/";
+                String cdnDir = source.replaceAll(".*://.*?(?=/)", cdnURL + pathPrefix + username);
+                FileSystem cdnRootFS = getRootFS(cdnDir);
+                Path from = new Path(source);
+                Path to = new Path(cdnDir);
+
+                if (!prereqSatisfied(cdnURL)) {
+                    warn("\"other_namenodes\" does not contain the CDN cluster address " + cdnURL);
+                    warn("The following steps will fail soon, and distcp will be skipped!");
+                    warn("Please add/append \"" + cdnURL + "\" to the \"other_namenodes\" attribute in your job specification.");
+                }
+
+                deleteDir(cdnRootFS, cdnDir);
+                runDistcp(from, to);
+                deleteDirOnExit(cdnRootFS, cdnDir);
+                addPermissionsToParents(cdnRootFS, cdnDir);
+                source = cdnDir;
+                info("Use CDN HDFS cluster: " + source);
+            } catch (Exception e) {
+                warn("An exception occurred during distcp: " + e.getMessage());
+                warn("Use original HDFS cluster: " + source);
+            }
+            info("####################################");
+            info("          End of Distcp");
+            info("####################################");
+        }
+    }
+
+    private void runDistcp(Path from, Path to) throws Exception {
+        info("source: " + from);
+        info("destination: " + to);
+        EmbeddedGobblin embeddedGobblin = new EmbeddedGobblinDistcp(from, to).mrMode();
+
+        // Used for global throttling"
+        embeddedGobblin.distributeJar("lib/*");
+
+        for (Map.Entry<String, String> entry : this.props.entrySet()) {
+            if (entry.getKey() != null && (entry.getKey()).startsWith("distcpConf.")) {
+                String key = (entry.getKey()).substring("distcpConf.".length());
+                embeddedGobblin.setConfiguration(key, entry.getValue());
+            }
+        }
+        JobExecutionResult result =  embeddedGobblin.run();
```

Comment:
sub dirs: same as source
parent dirs: depends on `push.cdn.readByOther` and `push.cdn.readByGroup`